R/settings.R

```diff
@@ -49,18 +38,41 @@ read_settings <- function(filename) {
   }
 }
 
+read_config_file <- function(config_file) {
+  if (is.null(config_file)) {
+    return(NULL)
+  }
+
+  config <- new.env()
+  if (endsWith(config_file, ".R")) {
+    load_config <- function(file) sys_source(file, config)
+    malformed <- function(e) {
+      stop("Malformed config file, ensure it is valid R syntax\n  ", conditionMessage(e), call. = FALSE)
+    }
+  } else {
+    load_config <- function(file) {
+      dcf_values <- read.dcf(file, all = TRUE)
+      for (setting in names(dcf_values)) {
+        tryCatch(
+          assign(setting, eval(str2lang(dcf_values[[setting]])), envir = config),
+          error = function(e) stop("Malformed config setting '", setting, "'\n  ", conditionMessage(e), call. = FALSE)
+        )
+      }
+    }
+    malformed <- function(e) {
+      stop("Malformed config file, ensure it ends in a newline\n  ", conditionMessage(e), call. = FALSE)
+    }
+  }
+  tryCatch(load_config(config_file), warning = malformed, error = malformed)
+  config
+}
+
 get_setting <- function(setting, config, defaults) {
   option <- getOption(paste(sep = ".", "lintr", setting))
   if (!is.null(option)) {
     option
   } else if (!is.null(config[[setting]])) {
-    malformed <- function(e) {
-      stop("Malformed config setting '", setting, "'\n  ", conditionMessage(e), call. = FALSE)
-    }
-    tryCatch(
-      eval(parse(text = config[[setting]])),
-      error = malformed
-    )
+    config[[setting]]
```

Comment:
We could use `%||%` here for a little less cyclocomp.

Reply:
way cleaner, nice.